Push reported again. Now with empty boolean support!

mzgoddard opened the pull request:
### Resolves

Fix  https://github.com/LLK/scratch-vm/issues/929

### Proposed Changes

Clear pushReported's optimization `justReported` member after reading it in `execute.js`. With `justReported` cleared, the value can't leak to other members of argValues to execute this stack frame's blockFunction with.

### Reason for Changes

Previously reported values leaked into boolean arguments that did not define a block. If those values were truthy that would change the execution behaviour when the scratch project's author intended it to be falsey.

### Test Coverage

There isn't a clear idiom how to test that the vm executes a block or procedure correctly. There are tests that test the specific block code in `src/blocks` and integration tests to see that the vm runs with those blocks. But not that the end or partial state of the integration test matches expected values. Anyways test coverage will be provided by another PR that will be easier to merge than this one as it'll just test the existing behaviour to ensure this specific regression doesn't occur in the future. (And this PR can provide the oppurtunity to discuss how to test this.)

mzgoddard committed:
Revert "Revert "Push reported""

This reverts commit 107adad647e63be9de90c8d8864d34609ef0701f.

mzgoddard committed:
Unset justReported after reading its value

Avoid justReported values leaking to other argValues by setting back to the
default null value.

mzgoddard commented:
One thought I have for testing that block execution happens as expect could be to connect to the runtime `SAY` event in integration tests. Says that start with `pass` or `fail` run `t.pass` or `t.fail`.

Using the procedure integration as an example:
![screen shot 2018-04-16 at 5 58 38 pm](https://user-images.githubusercontent.com/833298/38837472-177815f8-41a0-11e8-9d57-9aaf06c2c7d4.png)

Integrations that have more assertions could wrap that as they want with procedures if the integration isn't on procedure behaviour itself.
![screen shot 2018-04-16 at 5 56 26 pm](https://user-images.githubusercontent.com/833298/38837505-31480b1e-41a0-11e8-817b-911620ea2312.png)

I figure integration tests like this would be a new folder in the test folder and would have a js file automatically run all the sb2 and sb3 files in that folder perhaps.

kchadha reviewed (commented):
The new proposed changes (compared to the old push reported work) look good.

It did take a while to re-understand what's going on in execute. I'm wondering if there should be a high level comment explaining what's happening in execute in general. 

For example, I _think_ we know that there's never going to be an input and a field together, a block always has either/or, so it would help to have a description of 1) what's happening at a high level -- e.g. what information we need to look for/check for and 2) what is expected to be cached in any given run of execute (recursive or not).

kchadha commented:
As for testing, I do think we need to have some - are the blocks being interpreted as we think they should be - tests. Sort of going along with your 'say' idea, we could also have tests where stacks of blocks change one or more variables, and the test makes assertions about the value of the variable(s) at the end of running a stack. Of course, this would necessitate a test that ensures that variables are working correctly.

mzgoddard committed:
Comment execute's BlocksExecuteCache usage

Explain how BlocksExecuteCache helps speed up execute by storing values from
Blocks and one-time derived values on an object that is released when Blocks is
edited in the editor. The one time created cache object for a block then
simplifies the complexity of later javascript operations to use these stored
values.

mzgoddard commented:
@kchadha I added a comment https://github.com/LLK/scratch-vm/pull/1065/commits/19a32cd4898a9e9193362cd5a2fed8694280e543 for the BlocksExecuteCache area in execute and some other new comments.

mzgoddard commented:
I'm going to work on and make a separate PR for the test that could be reviewed and merged separately.

cwillisf commented on src/engine/execute.js:
```diff
@@ -98,32 +99,27 @@ const handleReport = function (
 };
 
+/**
+ * A convenienve constant to hide that the recursiveCall argument to execute is
```
s/convenienve/convenience/

cwillisf commented on src/engine/execute.js:
```diff
@@ -136,33 +132,97 @@ const execute = function (sequencer, thread) {
+
+        // Store the current shadow value if there is a shadow value.
+        blockCached._isShadowBlock = fieldKeys.length === 1 && Object.keys(inputs).length === 0;
+        blockCached._shadowValue = fieldKeys.length === 1 && fields[fieldKeys[0]].value;
```
Should this be:
```js
blockCached._shadowValue = blockCached._isShadowBlock && fields[fieldKeys[0]].value;
```
?

cwillisf commented on src/engine/execute.js:
```diff
@@ -173,37 +233,51 @@ const execute = function (sequencer, thread) {
+    // Add all fields on this block to the argValues. Some known fields may
+    // appear by themselves and can be set to argValues quicker by setting them
+    // explicitly.
+    if (blockCached._isFieldKnown) {
```
It looks like, with minor modification above, this could be simplified to something like:
```js
switch (blockCached._fieldKind) {
case foo.FIELD_VARIABLE:
    argValues.VARIABLE = blockCached._fieldVariable;
    break;
// similar cases for list & broadcast
default:
    for (const fieldName in fields) {
...
```

cwillisf commented on src/engine/execute.js:
```diff
@@ -212,7 +286,26 @@ const execute = function (sequencer, thread) {
+        let inputValue;
+        if (
+            currentStackFrame.waitingReporter === null
+        ) {
```
extra newlines above

cwillisf reviewed (changes requested):
I'm marking as "request changes" just for some weird formatting and a typo; the code looks ok to me. I mean, it does make me a bit nervous but I'm glad to see the updated tests as well as #1067. ;)

I did have a couple other suggestions & questions, but I don't think any of them are blocking.

mzgoddard commented:
@cwillisf I modified code in response to your review.
